### PR TITLE
#71 add searchbar component

### LINKED ISSUE #71: Add UI searchbar/input components
Add UI components for user input for searching product data.

## Expected Behavior

- Input displays placeholder on initial render.
- User can update input.

## Current Behavior

N/A.

## Possible Solution

See https://material-ui.com/components/text-fields.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #71.

Branched off #80 (suggest making your life easier by reviewing after #80 is merged).

## Description

- Adds `SearchBar` and dependencies.
- Adds stories for new components.
- Updates `MainPage` story to demo a table of entiites which can be filtered by `code` and `description`.

Next step:

- Wrap the functionality demoed in `MainPage` into its own organism (something like `EntityBrowser`). 
- Add Apollo client to implement functions for querying database.
- Add `EntityBrowser` to `MainPage` and implement `onSearch` callback to populate the entity data.
- Other "nice-to-haves": will probably need to add paging etc.
- MVP front-end for viewing data done, move onto authentication front-end!

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [x] Tests OK.

### CODE REVIEW

File: web/src/components/molecules/core/SearchBar.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    return (
+        <FlexContainer {...other}>
+            <FlexInputField input={input} icon={ClearIcon} onChange={onChange} onClick={onClear}/>
+            <FlexSearchButton onClick={onSearch}/>
+        </FlexContainer>
+    );
+};
+
+const styles = {
```
Comment (mark-prins) [inline web/src/components/molecules/core/SearchBar.tsx:30]: the styles would be better in a shared file

File: web/src/components/molecules/index.tsx [hunk 2]
```diff
@@ -1,2 +1,4 @@
 export { EntityTableHeader, EntityTableHeaderProps } from './entity';
```
Comment (mark-prins) [inline web/src/components/molecules/index.tsx:1]: already specified in the lower folder, could be ```
export * from './core';
export * from './entity';
```

File: web/src/components/atoms/core/index.tsx [hunk 3]
```diff
@@ -1,3 +1,10 @@
+export { Button, ButtonProps } from './Button';
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/index.tsx:1]: Hmm, thinking here is that any files exporting typescript components should be `tsx`. Yes? No?

File: web/src/components/atoms/core/SearchButton.tsx [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+import { Button, ButtonProps } from './Button';
+import { SearchIcon } from './SearchIcon';
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/SearchButton.tsx:3]: Yeah, my initial thinking was similar. There's a bit of intuition/flexibility, but in general a component containing another component isn't necessarily a molecule (I think classifying on that basis would also probably get us stuck quite quickly once we build up components, as there's only `atoms`, `molecules` and `organisms` at the component level, so we would probably run out of classifications!). 

Revisiting this one, though, I think it does look more like an atom!

File: web/src/components/atoms/index.tsx [hunk 5]
```diff
@@ -1,3 +1,10 @@
+export { Button, ButtonProps } from './core';
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/index.tsx:1]: Agreed!

File: web/src/components/index.tsx [hunk 6]
```diff
@@ -1,3 +1,14 @@
-export { EntityTable } from './organisms/entity/EntityTable';
-export { EntityTableRow } from './molecules/entity/EntityTableRow'; 
-export { EntityTableHeader } from './molecules/entity/EntityTableHeader'; 
\ No newline at end of file
+export { Button, ButtonProps } from './atoms';
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/index.tsx:1]: 👍.

File: web/src/components/atoms/core/SearchButton.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/SearchButton.tsx:3]: Yeah, my initial thinking was similar. There's a bit of intuition/flexibility, but in general a component containing another component isn't necessarily a molecule (I think classifying on that basis would also probably get us stuck quite quickly once we build up components, as there's only `atoms`, `molecules` and `organisms` at the component level, so we would probably run out of classifications!). 

In this case, it's still an atom in that it's a building block for a molecule, so I think it's probably the right classification (or at least better than having `SearchButton` and `SearchBar` both as molecules)?

File: web/src/components/molecules/core/InputField.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  ...other
+}) => {
+  const FlexContainer = React.useMemo(() => styled(({ ...props }) => <Container {...props}/>)(styles.flexContainer), []);
+  const FlexInput = React.useMemo(() => styled(({ ...props }) => <Input {...props}/>)(styles.flexInput), []);
+  const FlexButton = React.useMemo(() => styled(({ ...props }) => <Button {...props}/>)(styles.flexButton), []);
+  const ButtonIcon = React.useMemo(() => icon, [icon]);
+
+  return (
+    <FlexContainer {...other}>
+      <FlexInput disableUnderline value={input} onChange={onChange} />
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/molecules/core/InputField.tsx:31]: Hmm, can't actually remember why I disabled it! Should probably be a prop passed down from the parent organism.

File: web/src/components/molecules/core/SearchBar.tsx (hunk 1 above)
Comment (wlthomson) [reply web/src/components/molecules/core/SearchBar.tsx:30]: General design thinking here is to have shared styles in shared files (e.g. flex styles), but component specific styles in that component scope (in this case I think two components have the same 95/5 ratio of children, but IMO it's probably good to decouple them so we can style them separately if necessary/don't get style changes unexecptedly propagating around the place).

TLDR; I think this is OK as these styles are component specific so no need to import something that's only used in one file?

File: web/src/components/molecules/index.tsx (hunk 2 above)
Comment (wlthomson) [reply web/src/components/molecules/index.tsx:1]: 👍 👍 👍 .

File: web/src/components/atoms/core/SearchButton.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply web/src/components/atoms/core/SearchButton.tsx:3]: ok, thanks for clarifying. I'm perhaps being too literal in my reading of *atom* - good to know that it isn't necessarily a fundamental / non composed element. Just a small thing.
When we expand out the usage of atomic principles, we'd need to have some guidelines for people

File: web/src/components/atoms/core/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply web/src/components/atoms/core/index.tsx:1]: both are typescript - it's just that files containing react elements have to be `.tsx` in order to have the elements parsed. I seem to recall that there's a minor overhead in doing this, and therefore any files that are strictly javascript should be `.ts` ( same deal with `.jsx` and `.js` )
I may be wrong - and it might just be a convention that I've fallen into. I find it helpful to easily distinguish between react element files and vanilla script files.

File: web/src/components/molecules/core/SearchBar.tsx (hunk 1 above)
Comment (mark-prins) [reply web/src/components/molecules/core/SearchBar.tsx:30]: styles in react are a rich source of discussion!
I just commented on this one because the same style set was used in another component and it looked like a case of re-use. Creating a flex-container component perhaps.
Fine as you are.

File: web/src/components/atoms/core/SearchButton.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/SearchButton.tsx:3]: @mark-prins Hmm, now I'm second guessing myself... I think technically it is meant to be non-composed (can't be broken down into smaller elements), but then you end up with `SearchButton` and `SearchBar` being at the same classification level... but maybe that's OK?

Think I will go with your suggestion and keep atoms as non-divisible building blocks, makes organising components a bit less subjective!

File: web/src/components/molecules/index.tsx (hunk 2 above)
Comment (wlthomson) [reply web/src/components/molecules/index.tsx]: 61f74a738784c26382c8f0c2358f4dc60f48bc63

File: web/src/components/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/index.tsx:1]: 61f74a738784c26382c8f0c2358f4dc60f48bc63.

File: web/src/components/atoms/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/index.tsx]: 61f74a738784c26382c8f0c2358f4dc60f48bc63.

File: web/src/components/atoms/core/SearchButton.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/SearchButton.tsx]: 495db4e.

File: web/src/components/atoms/core/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (wlthomson) [reply web/src/components/atoms/core/index.tsx:1]: I agree that plain typescript should be `ts`, but I think it's a good convention to use `tsx` for files which export components (as they technically contain react elements)?
